### PR TITLE
Disallow directly reflecting closures

### PR DESCRIPTION
**Pull request by tbkka**
The Swift compiler uses various calling conventions for closures depending
on context.  This context disappears when you try to use Mirror to reflect
a closure-valued property, which can make the closure un-callable.

There doesn't seem to be any good solution to this, unless we
dictate that a closure's ABI is determined by it's type, which
we've so far been unwilling to guarantee.

So instead, this PR erases closure values during reflection,
making it impossible to directly reflect closures.
Note that the field still appears in the reflection list,
but the value is replaced with `()` (the empty tuple, aka Void).

```
struct A { var prop: () -> Int }
let a = A { return 7 }
let m = Mirror(reflecting: a)
let c = m.children.first!.value
let p = c as? () -> Int
// Before this PR, the cast above would succeed,
// but the call below could crash.
p()
```

Note however, that it _does_ work to put a closure in an `Any` or
other container, reflect that container, and then access the closure.
(In this case, the compiler has full visibility into all direct
accesses to the closure.)

```
struct B { var prop: Any }
let b = B(prop: { return 7 } as Any)
let m = Mirror(reflecting: b)
let c = m.children.first!.value
let p = c as? () -> Int // This always worked
p() // This always succeeded
```

Resolves rdar://59791962, SR-12268, SR-14362, #54696

### CODE REVIEW

**Comment by tbkka** [comment]
@swift-ci Please test

**Review by eeckstein — COMMENTED** [review]
nice!
I think it's worth documenting this in `Mirror.swift` (for `struct Mirror`) so that people can find this information in the official stdlib API documentation.

**Comment by tbkka** [comment]
I started documenting the default mirror construction, including the fact that it exposes private fields and the handling of closure-typed fields.  In the process, I realized that closure properties can also be accurately reflected via CustomMirror implementation, so I added a test for that.

**Comment by tbkka** [comment]
@swift-ci Please test

File: stdlib/public/runtime/ReflectionMirror.cpp [hunk 1]
```diff
@@ -416,8 +416,22 @@ getFieldAt(const Metadata *base, unsigned index) {
...
     typeInfo = result.getType();
   }
+  auto metadata = typeInfo.getMetadata();
 
-  auto fieldType = FieldType(typeInfo.getMetadata());
+  // We can't actually properly support closures via reflection.
+  // The compiler juggles several representations and runtime
+  // reflection loses enough context to make the closure un-callable.
+  // So we erase closure values in mirrors:
+  if (isa<FunctionTypeMetadata>(metadata)) {
```
[the remark this answers is not in the document]
Comment (tbkka) [reply stdlib/public/runtime/ReflectionMirror.cpp:425]: Hmmm....   C function pointers and blocks are called differently from each other, though, and reflection still loses that distinction, doesn't it?

[the remark this answers is not in the document]
Comment (jckarter) [reply stdlib/public/runtime/ReflectionMirror.cpp:425]: It shouldn't. `@convention(c)` and `@convention(block)` types are distinct in runtime type metadata.

**Comment by tbkka** [comment]
Since I last looked at this, a bunch of the reflection machinery is now being used in other places.  The change I was exploring here breaks a lot of tests for new reflection that I'll have to dig through before I can land this.

**Comment by tbkka** [comment]
Tentatively updated the tests so they all pass locally.  I still need to work through the C function pointer cases that Joe pointed out.

File: stdlib/public/runtime/ReflectionMirror.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (tbkka) [reply stdlib/public/runtime/ReflectionMirror.cpp]: Done.

**Pull request closed without merging**